template/template.go

```diff
@@ -296,3 +303,218 @@ func (e *executer) resolveOptions(options []ast.Option) (map[string]any, error)
 
 	return m, nil
 }
+
+func (e *executer) resolveMatcher(m ast.Matcher) error {
+	res, err := e.resolveSelector(m)
+	if err != nil {
+		return fmt.Errorf("resolve selector: %w", err)
+	}
+
+	pref := e.resolvePreferences(m, res)
+
+	filteredVariants := e.filterVariants(m, pref)
+
+	sortable := e.sortVariants(filteredVariants, pref)
+
+	err = e.selectBestVariant(sortable)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *executer) resolveSelector(matcher ast.Matcher) ([]any, error) {
+	res := make([]any, 0, len(matcher.MatchStatements))
+
+	for _, selectExpr := range matcher.MatchStatements {
+		var function ast.Function
+
+		switch annotation := selectExpr.Annotation.(type) {
+		case nil:
+			return nil, ErrMissingSelectorAnnotation
+		case ast.ReservedAnnotation, ast.PrivateUseAnnotation:
+			return nil, ErrUnsupportedExpression
+		case ast.Function:
+			function = annotation
+		}
+
+		registryF, ok := e.template.funcRegistry[function.Identifier.Name]
+		if !ok {
+			return nil, fmt.Errorf("%w '%s'", ErrUnknownFunction, function.Identifier.Name)
+		}
+
+		opts, err := e.resolveOptions(function.Options)
+		if err != nil {
+			return nil, fmt.Errorf("resolve options: %w", err)
+		}
+
+		input, err := e.resolveValue(selectExpr.Operand)
+		if err != nil {
+			return nil, fmt.Errorf("resolve value: %w", err)
+		}
+
+		rslt, err := registryF.Match(input, opts)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %s", ErrSelection, err.Error())
+		}
+
+		res = append(res, rslt)
+	}
+
+	return res, nil
+}
+
+func (e *executer) resolvePreferences(m ast.Matcher, res []any) [][]string {
+	// Step 2: Resolve Preferences
+	pref := make([][]string, 0, len(res))
+
+	for i := range res {
+		var keys []string
+
+		for _, variant := range m.Variants {
+			for _, vKey := range variant.Keys {
+				switch key := vKey.(type) {
+				case ast.CatchAllKey:
+					continue
+				case ast.NameLiteral:
+					keys = append(keys, string(key))
+				case ast.QuotedLiteral:
+					keys = append(keys, string(key))
```

Comment:
```suggestion
				case ast.NameLiteral, ast.QuotedLiteral:
					keys = append(keys, string(key))
```

In practice, we could use all three with `fmt.Sprint` but there is potentially a small performance penalty.

Reply:
we could not, Stringer for qutoed LIteral, prints string enclosed in |.
`|qutoed|`